core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java

```diff
@@ -269,17 +163,34 @@ public NativeImageBuildItem build(NativeConfig nativeConfig, NativeImageSourceJa
             errorReportLatch.await();
             int exitCode = process.waitFor();
             if (exitCode != 0) {
-                throw imageGenerationFailed(exitCode, command);
+                throw imageGenerationFailed(exitCode, buildCommand);
             }
             if (SystemUtils.IS_OS_WINDOWS && !(isContainerBuild)) {
                 //once image is generated it gets added .exe on Windows
                 executableName = executableName + ".exe";
             }
+
+            String symbols = String.format("%s.debug", executableName);
+            if (!isContainerBuild && !objcopyExists(env)) {
+                log.warn("objcopy executable not found in PATH. Debug symbols will not be separated from executable.");
+                log.warn("That will result in a larger native image with debug symbols embedded in it.");
+            } else {
+                assert !(isContainerBuild && containerCommand == null);
+                if (nativeConfig.debug.enabled) {
+                    objcopy(containerCommand, outputDir, "--only-keep-debug", executableName, symbols);
+                    objcopy(containerCommand, outputDir, String.format("--add-gnu-debuglink=%s", symbols), executableName);
+                }
+                // Strip debug symbols regardless, because the underlying JDK might contain them
+                objcopy(containerCommand, outputDir, "--strip-debug", executableName);
```

Comment:
Wouldn't this spin up a container 3 times? Ideally, we'd stitch them together so that we can get away with a one time container invocation. I.e. `podman run ... --entrypoint=/bin/bash <container> -c 'objcopy --only-keep-debug ... ; objcopy --add-gnu-debuglink=....; objcopy --strip-debug ...` or `podman run ... --entrypoint=/bin/bash <container> -c 'objcopy --strip-debug ...` depending on the `-g` option.

Reply:
It will, but that's fairly fast in comparison to the native-image build step. Keeping it this way we don't have to have two different code paths for container and host builds, thus keeping it simple and more readable.

Reply:
I haven't tested this myself, but apparently running containers on non-Linux isn't as quick. Perhaps mac and windows use-cases with building with the container should be tested.

Reply:
AFAIK Mac is almost as fast, and windows docker using WSL2 as the backend should also be pretty fast. I can do some testing on Windows, but unfortunately not on Mac.

Reply:
This maybe an optimization to do as a later step (once we know this actually is a problem?)

Reply:
FTR: I tested this on two laptops with similar specs.
I tested the "combined" (single docker container for all 3 processes) vs the "split" (1 container per process) approach and the results were:

| | Windows | Linux |
| -- | -- | -- |
| Combined | ~16s | ~0.8s |
| Split | 3.5 + 13 + 6.5 = ~23s | 0.5 + 0.5 + 0.6 = ~1.1s |

Although the benefit of using the "combined" approach is pretty clear, the overhead of the "split" approach (~7s) is still relatively low given that the image build took ~260s.